portable/GCC/ATmegaxxxx/readme.md

```diff
@@ -0,0 +1,84 @@
+
+For devices which can support __XRAM__ and have the __RAMPZ__ register, this register is also preserved during the context switch.
+
+<h3>Interrupt Nesting</h3>
+
+The ATmega family does not support interrupt nesting, having only one interrupt priority. This means that when the Scheduler is running, interrupts are normally disabled.
+
+When a very time critical process is running, based on microsecond timing generated by one of the Timers, it is important to re-enable interrupts as early as possible in processing a Yield. Fortunately, this is supported through the use of the `NO_BLOCK` decorator when defining the Interrupt Service Routine.
+
+The `NO_BLOCK` decorator will enable the global interrupt early in the handling of an ISR (in this case for the Scheduler), and enable interrupts to be nested. Using this method, I've been able to successfully implement an [Audio Synthesiser](https://feilipu.me/2015/06/02/goldilocks-analogue-synthesizer/) with less than 83 microseconds for each cycle, whilst still running the Scheduler to handle display and input.
+
+Using `NO_BLOCK` is optional, and should only be done if a critical Timer should interrupt the Scheduler.
+
+<h3>Heap Management</h3>
+
+Most users of FreeRTOS will choose to manage their own heap using one of the pre-allocated heap management algorithms, but for those that choose to use `heap_3.c`, the wrappered `malloc()` method, there is an issue that needs to be addressed.
+
+The avr-libc library assumes that the stack will always be above the heap, and does a check for this when responding to a `malloc()` request. This is not the case when Tasks are running, as their stack is located in the early allocated heap address ranges which will be below free heap memory, and so the `malloc()` request will fail even though heap space is available.
+
+To avoid this issue causing `pvPort_Malloc()` to failing, the user needs to issue this tuning statement BEFORE they use the heap, or use the `xTaskCreate()` API.
+
+```c
+if( __malloc_heap_end == 0 )
+    __malloc_heap_end = (char *)(RAMEND - __malloc_margin);
+```
+Unfortunately in the repository there is nowhere sensible to include this statement as it should be included early in the `main()` function.
+
+For devices which can support __XRAM__ the user will need to tune the location of stack and heap according to their own requirements.
+
+<h3>Supported Devices</h3>
+
+ATmega devices with __ENHANCED WDT__ Interrupt capability - will use WDT.
+
+ - ATmega8U2/16U2/32U2 -> 2kB RAM
+ - ATmega16U4/32U4 - Arduino Leonardo -> 2.5kB RAM
+ - ATmega48PB/88PB/168PB/328PB - Arduino Uno -> 2kB RAM
+ - ATmega164PA/324PA/644PA/1284P - Goldilocks -> __16kB RAM__
+ - ATmega324PB -> 2kB RAM
+ - ATmega640/1280/2560/1281/2561 - Arduino Mega -> __8kB RAM + XRAM__
+
+ATmega devices without enhanced __WDT__ Interrupt capability - will use Timer0.
+
+ - ATmega8A/16A/32A/64A/128A -> 4kB RAM
+ - ATmega808/809/1608/1609/3208/3209/4808/4809 -> 6kB RAM
```

Comment:
ATmega808/809/1608/1609/3208/3209/4808/4809 series is comparatively new. 
It doesn't have a dedicated 8 bit timer Timer0. It has only 16 bit timers named TCA and TCB. These can be used in split-mode to create 8 bit timers if required. Also, TCA and TCB have completely different set of registers and bitmaps than that of legacy mega devices such as mega48PB/mega328PB. I think, with Timer0, this PR can't serve ATmega808/809/1608/1609/3208/3209/4808/4809. What do you think?

Reply:
Grrr. Yes. It is a totally different beast.

Looking at the data sheet the best option seems to use the RTC with the internal 32kHz Oscillator in the Periodic Interrupt Timer (PIT) mode as the System Tick.

_Using the same clock source as the RTC function, the PIT can request an interrupt or trigger an output event on every nth clock period. n can be selected from {4, 8, 16,.. 32768} for interrupts, and from {64, 128, 256,... 8192} for events._

The ATmega808/809/1608/1609/3208/3209/4808/4809 WDT doesn't have any interrupt capability, and I'd picked up that it couldn't be used. But I hadn't noticed the completely reimagined Timer solution.

I don't have any of ATmega808/809/1608/1609/3208/3209/4808/4809, so I'd suggest to work this out later.

Reply:
Further, I'm quite sure that avr-libc doesn't support this series of devices, as [2.0.0 was released in 2016](https://savannah.nongnu.org/forum/forum.php?forum_id=8460). Without avr-libc, everything here is irrelevant.

The only option is to make a note that the ATmega808/809/1608/1609/3208/3209/4808/4809 series is unsupported.

Reply:
I think, megaAVR-0 series is of architecture **avr8x** - which is the same as xmega and few tiny devices. This is supported by avr-libc since the architecture was already existing.
I have one 4808. May be I can try after this got merged.

Reply:
Ok. That makes sense. I've not seen that avr8.x architecture before.

Then it would be best to handle support as part of an __ATxmega__ port.

Something to add later, or separately.

Reply:
Another solution I could think of is probably making ```prvSetupTimerInterrupt()``` in port.c external. We could provide example implementations in various projects, e.g. 8-bit/16-bit timer (as default probably), watchdog as a second clock source (for security initiatives), and RTC (for use cases concerning about clock skew).  Users could determine which one(s) to use in their application scenario. 

This may make sense, since all the timers we are talking about here are not part of the MCU core but rather peripheral. (In contrast to ARM, SysTick is provided in MCU core.) This will also make port cleaner...

Reply:
Agreed. As noted in another comment, putting the TimerN configuration into a separate file is what I do in practice. I usually hold all the timer configurations in a separate file called [`FreeRTOSBoardDefs.h`](https://github.com/feilipu/avrfreertos/blob/master/freeRTOS10xx/include/FreeRTOSBoardDefs.h), or [`FreeRTOSVariant.h`](https://github.com/feilipu/Arduino_FreeRTOS_Library/blob/master/src/FreeRTOSVariant.h) for ArduinoFreeRTOS.

But, this is unrelated to the OP comment. Support for this new ATmega808/809/1608/1609/3208/3209/4808/4809 `avr8x` family is going to be difficult as the tool set required is not yet fully developed. Something that can be looked at as a future addition, imho.

Reply:
As far as I know, these devices are supported using a device spec file from DFP.
These devices are supproted in Atmel START and MCC for public. In fact, few IoT [examples](https://start.atmel.com/#examples/avr/iot/wg) are also available in 4808 which consumes almost 90% of memory resources. Anyway, I can check and confirm if there is any limitation.

Reply:
I have checked the `portmacro.h` and `port.c` file from the v10.0.0 files included in a ATmega4809 platform files.
 - `portmacro.h` - no relevant differences, but uses some depreciated gcc grammar.
 - `port.c` - removed optional debugging bytes in `pxPortInitialiseStack`, only standard 2-byte stack management, and some depreciated gcc grammar.

I think, if the `prvSetupTimerInterrupt()` is kept as an external or weak, along with the `portSCHEDULER_ISR` definition being weak, there are is nothing preventing support for these working.

I can add the small differences if desired. @yuhui-zheng please advise.